feat: add `fft/base/fftpack/rfftf` (#12002)
stdlib-js/stdlib: lib/node_modules/@stdlib/fft/base/fftpack/rfftf/README.md

<!--

@license Apache-2.0

Copyright (c) 2026 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# rfftf

> > Compute the forward real-valued fast Fourier transform (FFT).

Reviewer: ```suggestion
> Compute the forward discrete Fourier transform (DFT) of a real-valued sequence.
```

<!-- Section to include introductory text. Make sure to keep an empty line after the intro `section` element and another before the `/section` close. -->

<section class="intro">

</section>

<!-- /.intro -->

<!-- Package usage documentation. -->

<section class="usage">

## Usage

```javascript
var rfftf = require( '@stdlib/fft/base/fftpack/rfftf' );
```

#### rfftf( N, r, strideR, offsetR, workspace, strideW, offsetW )

> Computes the forward real-valued fast Fourier transform (FFT).

Reviewer: ```suggestion
Computes the forward discrete Fourier transform (DFT) of a real-valued sequence.
```

```javascript
var Float64Array = require( '@stdlib/array/float64' );
var rffti = require( '@stdlib/fft/base/fftpack/rffti' );

var N = 4;
var workspace = new Float64Array( ( 2*N ) + 34 );

rffti( N, workspace, 1, 0 );

var r = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

rfftf( N, r, 1, 0, workspace, 1, 0 );

// r => <Float64Array>[ 10.0, -2.0, 2.0, -2.0 ]
```

The function accepts the following arguments:

> -   **N**: length of the sequence to transform.

Reviewer: ```suggestion
-   **N**: length of the sequence to transform. The function is most efficient when this value is a product of small prime numbers.
```

-   **r**: input array.
-   **strideR**: stride length for `r`.
-   **offsetR**: starting index for `r`.
-   **workspace**: workspace array containing pre-computed values.
-   **strideW**: stride length for `workspace`.
-   **offsetW**: starting index for `workspace`.

</section>

<!-- /.usage -->

<!-- Package usage notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="notes">

## Notes

-   Before calling this function, initialize the workspace by calling [`rffti`][@stdlib/fft/base/fftpack/rffti] with the same sequence length and workspace layout.

> -   The function performs the transform in-place and returns the same input array reference.

Reviewer: ```suggestion
-   The function performs the transform in-place (i.e., the input array is **mutated**).
```

-   For `N = 4`, the output

    ```text
    [ 10.0, -2.0, 2.0, -2.0 ]
    ```

    corresponds to a zero-frequency term `10.0`, a complex coefficient `-2.0 + 2.0i` at frequency `1`, and a Nyquist term `-2.0`.

> -   If `N` equals `1`, the function returns early without modifying the input, as a single data point is its own Fourier transform. In that case, the function still returns the same input array reference.

Reviewer: ```suggestion
-   If `N` equals `1`, the function returns early without modifying the input, as a single data point is its own Fourier transform. 

-   This transform is unnormalized as a call to this function followed by a call performing a [backward transform][@stdlib/fft/base/fftpack/rfftb] will multiply the input array by `N`.
```

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var zeros = require( '@stdlib/array/zeros' );
var discreteUniform = require( '@stdlib/random/array/discrete-uniform' );
var rffti = require( '@stdlib/fft/base/fftpack/rffti' );
var rfftf = require( '@stdlib/fft/base/fftpack/rfftf' );

var N = 4;
var opts = {
    'dtype': 'float64'
};
var r = discreteUniform( N, -10, 10, opts );
var workspace = zeros( ( 2*N ) + 34 );

console.log( r );

rffti( N, workspace, 1, 0 );
rfftf( N, r, 1, 0, workspace, 1, 0 );

console.log( r );
```

</section>

<!-- /.examples -->

<!-- Section to include cited references. If references are included, add a horizontal rule *before* the section. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="references">

</section>

<!-- /.references -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

> [@stdlib/fft/base/fftpack/rffti]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/fft/base/fftpack/rffti

Reviewer: ```suggestion
[@stdlib/fft/base/fftpack/rffti]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/fft/base/fftpack/rffti

[@stdlib/fft/base/fftpack/rfftb]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/fft/base/fftpack/rfftb
```
